--- PyGame_SDL2/setup.py.orig	2025-05-21 11:48:43.701351573 -0700
+++ PyGame_SDL2/setup.py	2025-05-21 11:50:04.758374404 -0700
@@ -162,19 +162,11 @@
         os.unlink(os.path.join(os.path.dirname(__file__), "src", "pygame_sdl2", i))
 
     if headers:
-        import pathlib
-
-        virtual_env = os.environ.get("VIRTUAL_ENV", None)
-
-        if virtual_env:
-            headers_dir = pathlib.Path(virtual_env) / "include" / "pygame_sdl2"
-        else:
-            headers_dir = pathlib.Path(sysconfig.get_paths()['include']) / "pygame_sdl2"
-
-        headers_dir.mkdir(parents=True, exist_ok=True)
+        headers_dir = os.path.join(os.environ['FLATPAK_DEST'], 'include', 'pygame_sdl2')
+        try:
+            os.makedirs(headers_dir)
+        except:
+            pass
 
         for header in headers:
-            srcpath = pathlib.Path(header)
-            dstpath = headers_dir / srcpath.name
-
-            shutil.copy(srcpath, dstpath)
+            shutil.copy(header, os.path.join(headers_dir, os.path.basename(header)))
