
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Interpretability: Inferring importances for GO and genes &#8212; NetworkVI 1.0.0 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../_static/documentation_options.js?v=f9c1cefe"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'notebooks/go_analysis';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Interpretability: Inferring GO-specific covariate attention values" href="go_specific_covariate_attention.html" />
    <link rel="prev" title="Mosaic integration" href="mosaic_integration.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">NetworkVI 1.0.0 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../api.html">API</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference/networkvi.model.NETWORKVI.html">networkvi.model.NETWORKVI</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/networkvi.nn.GeneLayers.html">networkvi.nn.GeneLayers</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference/networkvi.nn.GOLayers.html">networkvi.nn.GOLayers</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../tutorials.html">Tutorials</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="paired_integration_and_query_mapping.html">Paired integration and query-to-reference mapping</a></li>
<li class="toctree-l2"><a class="reference internal" href="mosaic_integration.html">Mosaic integration</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Interpretability: Inferring importances for GO and genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="go_specific_covariate_attention.html">Interpretability: Inferring GO-specific covariate attention values</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../references.html">References</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../_sources/notebooks/go_analysis.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Interpretability: Inferring importances for GO and genes</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-loading">Data loading</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-setup">Data setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#model-setup-and-training">Model setup and training</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#networkvi">NetworkVI</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretability">Interpretability</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="interpretability-inferring-importances-for-go-and-genes">
<h1>Interpretability: Inferring importances for GO and genes<a class="headerlink" href="#interpretability-inferring-importances-for-go-and-genes" title="Link to this heading">#</a></h1>
<p>NetworkVI is a Variational Autoencoder for the paired and mosaic integration of multimodal single-cell data. Each single-cell modality is learned in individual encoders; the latent spaces are then aligned for a joint representation. In this notebook we demonstrate how to use the inherent interpretability of NetworkVI to infer importances for GO terms and genes. We use a publically available CITE-seq dataset created by <span id="id1">[<a class="reference internal" href="../references.html#id4" title="Malte Luecken, Daniel Burkhardt, Robrecht Cannoodt, Christopher Lance, Aditi Agrawal, Hananeh Aliee, Ann Chen, Louise Deconinck, Angela Detweiler, Alejandro Granados, Shelly Huynh, Laura Isacco, Yang Kim, Dominik Klein, BONY DE KUMAR, Sunil Kuppasani, Heiko Lickert, Aaron McGeever, Joaquin Melgarejo, Honey Mekonen, Maurizio Morri, Michaela Müller, Norma Neff, Sheryl Paul, Bastian Rieck, Kaylie Schneider, Scott Steelman, Michael Sterr, Daniel Treacy, Alexander Tong, Alexandra-Chloe Villani, Guilin Wang, Jia Yan, Ce Zhang, Angela Pisco, Smita Krishnaswamy, Fabian Theis, and Jonathan M Bloom. A sandbox for prediction and integration of dna, rna, and proteins in single cells. In J. Vanschoren and S. Yeung, editors, Proceedings of the Neural Information Processing Systems Track on Datasets and Benchmarks, volume 1. 2021. URL: https://datasets-benchmarks-proceedings.neurips.cc/paper_files/paper/2021/file/158f3069a435b314a80bdcb024f8e422-Paper-round2.pdf.">Luecken <em>et al.</em>, 2021</a>]</span>.</p>
<p>If you use NetworkVI, please consider citing:</p>
<p>Arnoldt, L., Upmeier zu Belzen, J., Herrmann, L., Nguyen, K., Theis, F.J., Wild, B. , Eils, R., “Biologically Guided Variational Inference for Interpretable Multimodal Single-Cell Integration”, bioRxiv, June 2025.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">os</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>

<span class="n">IN_COLAB</span> <span class="o">=</span> <span class="s2">&quot;google.colab&quot;</span> <span class="ow">in</span> <span class="n">sys</span><span class="o">.</span><span class="n">modules</span>

<span class="k">if</span> <span class="n">IN_COLAB</span><span class="p">:</span>
    <span class="o">!</span>pip<span class="w"> </span>install<span class="w"> </span>networkvi
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">requests</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sys</span>
<span class="n">sys</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;../../../src&quot;</span><span class="p">)</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">networkvi</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">networkvi.model</span><span class="w"> </span><span class="kn">import</span> <span class="n">NETWORKVI</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">scanpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sc</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">anndata</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">ad</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib.pyplot</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">plt</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.ensemble</span><span class="w"> </span><span class="kn">import</span> <span class="n">RandomForestClassifier</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">sklearn</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.preprocessing</span><span class="w"> </span><span class="kn">import</span> <span class="n">normalize</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">seaborn</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">sns</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">goatools.obo_parser</span><span class="w"> </span><span class="kn">import</span> <span class="n">GODag</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">shutil</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%%capture</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">warnings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">sklearn.exceptions</span><span class="w"> </span><span class="kn">import</span> <span class="n">ConvergenceWarning</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="n">ConvergenceWarning</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">r</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;http://purl.obolibrary.org/obo/go/go-basic.obo&quot;</span><span class="p">,</span> <span class="n">allow_redirects</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;../../../resources/&quot;</span><span class="p">,</span> <span class="s2">&quot;go-basic.obo&quot;</span><span class="p">),</span> <span class="s1">&#39;wb&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">r</span><span class="o">.</span><span class="n">content</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>31321250
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="s2">&quot;../../../gene_interactions.csv&quot;</span><span class="p">):</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">gdown</span>
    <span class="n">gdown</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s2">&quot;https://drive.google.com/uc?export=download&amp;id=1MwAuqw2JVl6L9xfsWGfUH7rB02LsQ_Vv&quot;</span><span class="p">,</span> <span class="s2">&quot;../../../gene_interactions.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<section id="data-loading">
<h2>Data loading<a class="headerlink" href="#data-loading" title="Link to this heading">#</a></h2>
<p>We provide a subsetted and preprocessed version of the CITE-seq and Multiome dataset for this tutorial. We filtered low qualty cells, lowly expressed features and performed subsampling to 5000 cells and feature subselection to 4000 highly variable genes and 20000 highly variable ATAC peaks. NetworkVI accepts raw counts data for all modalities. For details on the preprocessing please refer to the manuscript and the code in the <a class="reference external" href="https://github.com/LArnoldt/networkvi_reproducibility">NetworkVI reproducibility repository</a>. In the repository we also provide scripts for downloading and processing the datasets used in the manuscript. The original CITE-seq dataset contains 90261 bone marrow mononuclear cells inferred at 4 different sites with 13953 genes and 143 proteins.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">rna_cite_path</span> <span class="o">=</span> <span class="s2">&quot;neurips2021_cite_bmmc_luecken2021.h5ad&quot;</span>

<span class="k">try</span><span class="p">:</span>
    <span class="n">rna_cite</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">rna_cite_path</span><span class="p">)</span>
<span class="k">except</span> <span class="ne">OSError</span><span class="p">:</span>
    <span class="kn">import</span><span class="w"> </span><span class="nn">gdown</span>

    <span class="n">gdown</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s2">&quot;https://drive.google.com/uc?export=download&amp;id=1A9o8wZgWS6udFMXJ-ybXre1dCcCdsKt5&quot;</span><span class="p">)</span>
    <span class="n">rna_cite</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">rna_cite_path</span><span class="p">)</span>

<span class="n">rna_cite</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 5000 × 4000
    obs: &#39;cell_type&#39;, &#39;Site&#39;, &#39;DonorID&#39;, &#39;DonorSmoker&#39;, &#39;DonorBMI&#39;, &#39;DonorAge&#39;
    var: &#39;gene_stable_id&#39;, &#39;modality&#39;
    uns: &#39;protein_expression&#39;
    obsm: &#39;protein_expression&#39;
</pre></div>
</div>
</div>
</div>
</section>
<section id="data-setup">
<h2>Data setup<a class="headerlink" href="#data-setup" title="Link to this heading">#</a></h2>
<p>To perform query-to-reference mapping, we select the cells inferred at one of the four different sites as the query dataset. We train NetworkVI on the rest of the dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">rna_cite</span>
<span class="n">query</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Site&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;site1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;Site&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;site1&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>We register the fields and setup the anndata with a convenience function as shown below. NetworkVI requires as MultiVi to register the modality as the batch_key. Further batch information can be registered as categorical_covariate_keys.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">NETWORKVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;Site&quot;</span><span class="p">,</span>
    <span class="n">protein_expression_obsm_key</span><span class="o">=</span><span class="s2">&quot;protein_expression&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="model-setup-and-training">
<h2>Model setup and training<a class="headerlink" href="#model-setup-and-training" title="Link to this heading">#</a></h2>
<p>We demonstrate the setup and training of NetworkVI, which makes use of the Gene Ontology, as the encoder structure. Below we define a convenience function to get the latent representation and to visualize the latent space. NetworkVI requires as MultiVI the definition of the number of features per modality (n_genes, n_proteins, n_regions).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">get_plot_latent_representation</span><span class="p">(</span><span class="n">vae</span><span class="p">,</span> <span class="n">adata</span><span class="p">):</span>
    <span class="n">latent_representation</span> <span class="o">=</span> <span class="n">vae</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">(</span><span class="n">modality</span><span class="o">=</span><span class="s2">&quot;joint&quot;</span><span class="p">)</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_NetworkVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">latent_representation</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_NetworkVI&quot;</span><span class="p">)</span>
    <span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_components</span><span class="o">=</span><span class="mi">2</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">adata</span>
</pre></div>
</div>
</div>
</div>
<section id="networkvi">
<h3>NetworkVI<a class="headerlink" href="#networkvi" title="Link to this heading">#</a></h3>
<p>We select NetworkVI by setting layers_encoder_type to “go”. The mapping of the features to the gene layer and the Gene Ontology requires ENSEMBL-IDs, which we provide in ensembl_ids_genes and ensembl_ids_proteins. We also provide the Gene Ontology and a gene-to-GO mapping file in obo_file and map_ensembl_go. Please don’t use the provided mapping file in production since it’s an artifically subsampled version of a real mapping file. Please refer to the scripts provided in https://github.com/Larnoldt/networkvi_reproducibility or https://geneontology.org/docs/go-annotation-file-gaf-format-2.1/ to download full mapping files.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vae</span> <span class="o">=</span> <span class="n">NETWORKVI</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">n_genes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;modality&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Gene Expression&quot;</span><span class="p">]),</span>
    <span class="n">n_proteins</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;protein_expression&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>
    <span class="n">ensembl_ids_genes</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">var</span><span class="p">[</span><span class="s2">&quot;modality&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;Gene Expression&quot;</span><span class="p">][</span><span class="s2">&quot;gene_stable_id&quot;</span><span class="p">]),</span>
    <span class="n">ensembl_ids_proteins</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;protein_expression&quot;</span><span class="p">][</span><span class="s2">&quot;var&quot;</span><span class="p">][</span><span class="s2">&quot;gene_stable_id&quot;</span><span class="p">]),</span>
    <span class="n">gene_layer_interaction_source</span><span class="o">=</span><span class="s2">&quot;../../../gene_interactions.csv&quot;</span><span class="p">,</span>
    <span class="n">expression_gene_layer_type</span><span class="o">=</span><span class="s2">&quot;interaction&quot;</span><span class="p">,</span>
    <span class="n">protein_gene_layer_type</span><span class="o">=</span><span class="s2">&quot;interaction&quot;</span><span class="p">,</span>
    <span class="n">obo_file</span><span class="o">=</span><span class="s2">&quot;../../../resources/go-basic.obo&quot;</span><span class="p">,</span>
    <span class="n">map_ensembl_go</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;../../../resources/ensembl2go.gaf&quot;</span><span class="p">],</span>
    <span class="n">layers_encoder_type</span><span class="o">=</span><span class="s2">&quot;go&quot;</span><span class="p">,</span>
    <span class="n">encode_covariates</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">deeply_inject_covariates</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">fully_paired</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">standard_gene_size</span><span class="o">=</span><span class="mi">5</span><span class="p">,</span>
    <span class="n">standard_go_size</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
<span class="p">)</span>

<span class="n">vae</span><span class="o">.</span><span class="n">train</span><span class="p">(</span><span class="n">max_epochs</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">adversarial_mixing</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">save_best</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated block structure with len 4000 (0.00025 blocks).
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated block structure with len 7342 (0.000458875 blocks).
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>../../../resources/go-basic.obo: fmt(1.2) rel(2025-03-16) 43,544 Terms; optional_attrs(relationship)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>../../../resources/go-basic.obo: fmt(1.2) rel(2025-03-16) 43,544 Terms; optional_attrs(relationship)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>_biological_process_go-basic_ensembl2go.v2_processed.pkl
Using attention in GO model at level -1.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated block structure with len 4565 (0.07608333333333334 blocks).
Generated block structure with len 4565 (0.07608333333333334 blocks).
Using attention in GO model at level 1.
Generated block structure with len 18 (0.13333333333333333 blocks).
Generated block structure with len 18 (0.13333333333333333 blocks).
Generated block structure with len 297 (0.00825 blocks).
Generated block structure with len 297 (0.00825 blocks).
Using attention in GO model at level 0.
Generated block structure with len 1 (0.06666666666666667 blocks).
Generated block structure with len 1 (0.06666666666666667 blocks).
Generated block structure with len 8 (0.8888888888888888 blocks).
Generated block structure with len 8 (0.8888888888888888 blocks).
Generated block structure with len 134 (0.0078125 blocks).
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Generated block structure with len 154 (0.0093994140625 blocks).
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>../../../resources/go-basic.obo: fmt(1.2) rel(2025-03-16) 43,544 Terms; optional_attrs(relationship)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>../../../resources/go-basic.obo: fmt(1.2) rel(2025-03-16) 43,544 Terms; optional_attrs(relationship)
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Using attention in GO model at level -1.
Generated block structure with len 182 (0.109375 blocks).
Generated block structure with len 182 (0.109375 blocks).
Using attention in GO model at level 1.
Generated block structure with len 16 (0.15384615384615385 blocks).
Generated block structure with len 16 (0.15384615384615385 blocks).
Generated block structure with len 27 (0.0263671875 blocks).
Generated block structure with len 27 (0.0263671875 blocks).
Using attention in GO model at level 0.
Generated block structure with len 1 (0.07692307692307693 blocks).
Generated block structure with len 1 (0.07692307692307693 blocks).
Generated block structure with len 7 (0.875 blocks).
Generated block structure with len 7 (0.875 blocks).
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True (cuda), used: True
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>TPU available: False, using: 0 TPU cores
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>HPU available: False, using: 0 HPUs
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>You are using a CUDA device (&#39;NVIDIA A100-PCIE-40GB&#39;) that has Tensor Cores. To properly utilize them, you should set `torch.set_float32_matmul_precision(&#39;medium&#39; | &#39;high&#39;)` which will trade-off precision for performance. For more details, read https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html#torch.set_float32_matmul_precision
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Training:   0%|                                                                                                                           | 0/10 [00:00&lt;?, ?it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 1/10:   0%|                                                                                                                         | 0/10 [00:00&lt;?, ?it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 1/10:  10%|███████████▎                                                                                                     | 1/10 [00:02&lt;00:18,  2.06s/it]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 1/10:  10%|█████▎                                               | 1/10 [00:02&lt;00:18,  2.06s/it, v_num=1, train_loss_step=5.09e+3, train_loss_epoch=1.03e+4]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 2/10:  10%|█████▎                                               | 1/10 [00:02&lt;00:18,  2.06s/it, v_num=1, train_loss_step=5.09e+3, train_loss_epoch=1.03e+4]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 2/10:  20%|██████████▌                                          | 2/10 [00:03&lt;00:12,  1.62s/it, v_num=1, train_loss_step=5.09e+3, train_loss_epoch=1.03e+4]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 2/10:  20%|██████████▊                                           | 2/10 [00:03&lt;00:12,  1.62s/it, v_num=1, train_loss_step=4.18e+3, train_loss_epoch=8.9e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 3/10:  20%|██████████▊                                           | 2/10 [00:03&lt;00:12,  1.62s/it, v_num=1, train_loss_step=4.18e+3, train_loss_epoch=8.9e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 3/10:  30%|████████████████▏                                     | 3/10 [00:04&lt;00:10,  1.48s/it, v_num=1, train_loss_step=4.18e+3, train_loss_epoch=8.9e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 3/10:  30%|███████████████▉                                     | 3/10 [00:04&lt;00:10,  1.48s/it, v_num=1, train_loss_step=5.83e+3, train_loss_epoch=7.96e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 4/10:  30%|███████████████▉                                     | 3/10 [00:04&lt;00:10,  1.48s/it, v_num=1, train_loss_step=5.83e+3, train_loss_epoch=7.96e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 4/10:  40%|█████████████████████▏                               | 4/10 [00:05&lt;00:08,  1.41s/it, v_num=1, train_loss_step=5.83e+3, train_loss_epoch=7.96e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 4/10:  40%|█████████████████████▏                               | 4/10 [00:05&lt;00:08,  1.41s/it, v_num=1, train_loss_step=9.61e+3, train_loss_epoch=7.22e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 5/10:  40%|█████████████████████▏                               | 4/10 [00:05&lt;00:08,  1.41s/it, v_num=1, train_loss_step=9.61e+3, train_loss_epoch=7.22e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 5/10:  50%|██████████████████████████▌                          | 5/10 [00:07&lt;00:06,  1.37s/it, v_num=1, train_loss_step=9.61e+3, train_loss_epoch=7.22e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 5/10:  50%|███████████████████████████                           | 5/10 [00:07&lt;00:06,  1.37s/it, v_num=1, train_loss_step=4.53e+3, train_loss_epoch=6.6e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 6/10:  50%|███████████████████████████                           | 5/10 [00:07&lt;00:06,  1.37s/it, v_num=1, train_loss_step=4.53e+3, train_loss_epoch=6.6e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 6/10:  60%|████████████████████████████████▍                     | 6/10 [00:08&lt;00:05,  1.35s/it, v_num=1, train_loss_step=4.53e+3, train_loss_epoch=6.6e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 6/10:  60%|███████████████████████████████▊                     | 6/10 [00:08&lt;00:05,  1.35s/it, v_num=1, train_loss_step=3.56e+3, train_loss_epoch=6.05e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 7/10:  60%|███████████████████████████████▊                     | 6/10 [00:08&lt;00:05,  1.35s/it, v_num=1, train_loss_step=3.56e+3, train_loss_epoch=6.05e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 7/10:  70%|█████████████████████████████████████                | 7/10 [00:09&lt;00:04,  1.34s/it, v_num=1, train_loss_step=3.56e+3, train_loss_epoch=6.05e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 7/10:  70%|█████████████████████████████████████                | 7/10 [00:09&lt;00:04,  1.34s/it, v_num=1, train_loss_step=8.27e+3, train_loss_epoch=5.59e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 8/10:  70%|█████████████████████████████████████                | 7/10 [00:09&lt;00:04,  1.34s/it, v_num=1, train_loss_step=8.27e+3, train_loss_epoch=5.59e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 8/10:  80%|██████████████████████████████████████████▍          | 8/10 [00:11&lt;00:02,  1.33s/it, v_num=1, train_loss_step=8.27e+3, train_loss_epoch=5.59e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 8/10:  80%|██████████████████████████████████████████▍          | 8/10 [00:11&lt;00:02,  1.33s/it, v_num=1, train_loss_step=4.08e+3, train_loss_epoch=5.16e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 9/10:  80%|██████████████████████████████████████████▍          | 8/10 [00:11&lt;00:02,  1.33s/it, v_num=1, train_loss_step=4.08e+3, train_loss_epoch=5.16e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 9/10:  90%|███████████████████████████████████████████████▋     | 9/10 [00:12&lt;00:01,  1.32s/it, v_num=1, train_loss_step=4.08e+3, train_loss_epoch=5.16e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 9/10:  90%|████████████████████████████████████████████████▌     | 9/10 [00:12&lt;00:01,  1.32s/it, v_num=1, train_loss_step=7.8e+3, train_loss_epoch=4.82e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10:  90%|███████████████████████████████████████████████▋     | 9/10 [00:12&lt;00:01,  1.32s/it, v_num=1, train_loss_step=7.8e+3, train_loss_epoch=4.82e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10: 100%|████████████████████████████████████████████████████| 10/10 [00:13&lt;00:00,  1.32s/it, v_num=1, train_loss_step=7.8e+3, train_loss_epoch=4.82e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10: 100%|███████████████████████████████████████████████████| 10/10 [00:13&lt;00:00,  1.32s/it, v_num=1, train_loss_step=4.56e+3, train_loss_epoch=4.51e+3]
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>`Trainer.fit` stopped: `max_epochs=10` reached.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 10/10: 100%|███████████████████████████████████████████████████| 10/10 [00:13&lt;00:00,  1.38s/it, v_num=1, train_loss_step=4.56e+3, train_loss_epoch=4.51e+3]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<p>We can visualize the generated latent space:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">get_plot_latent_representation</span><span class="p">(</span><span class="n">vae</span><span class="p">,</span> <span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="s2">&quot;Site&quot;</span><span class="p">],</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/ffedefb72885c849fd002e17aeea63e96e109afe0e90e83c46933e236a58eb8d.png" src="../_images/ffedefb72885c849fd002e17aeea63e96e109afe0e90e83c46933e236a58eb8d.png" />
</div>
</div>
</section>
</section>
<section id="interpretability">
<h2>Interpretability<a class="headerlink" href="#interpretability" title="Link to this heading">#</a></h2>
<p>By fitting a Logistic Regression on the activations of the nodes representing a GO term (calc_go_terms), a gene (calc_genes) or a gene_group (calc_gene_groups) we may infer the importance of this node for the performance of the neural network as described previously by <span id="id2">[<a class="reference internal" href="../references.html#id6" title="Jianzhu Ma, Michael Ku Yu, Samson Fong, Keiichiro Ono, Eric Sage, Barry Demchak, Roded Sharan, and Trey Ideker. Using deep learning to model the hierarchical structure and function of a cell. Nature Methods, 15(4):290–298, 2018. URL: https://doi.org/10.1038/nmeth.4627, doi:10.1038/nmeth.4627.">Ma <em>et al.</em>, 2018</a>]</span>. Here, we demonstrate how to utilize the get_go_gene_importances and get_perturbation_go_gene_importances functions to infer the importances for the original dataset and for perturbed datasets in a sensitivity analysis. The importance is calculated in standard mode for each label against the rest. Using labels_selection, labels_selection_groups, restrict_by_column_key_activations_mask, restrict_by_column_values_activations_mask and comparison this behaviour may be modified. Please refer to the API documentation.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">obodag</span> <span class="o">=</span> <span class="n">GODag</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="s2">&quot;../../../resources/&quot;</span><span class="p">,</span> <span class="s2">&quot;go-basic.obo&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>../../../resources/go-basic.obo: fmt(1.2) rel(2025-03-16) 43,544 Terms
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">vae</span><span class="o">.</span><span class="n">get_go_gene_importances</span><span class="p">(</span><span class="n">labels_column</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
                                <span class="n">results_dir</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;tutorial/calc_go_terms&quot;</span><span class="p">,</span>
                                <span class="n">shuffle_set_split</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">batch_size</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
                                <span class="n">calc_go_terms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                <span class="n">calc_genes</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">calc_gene_groups</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                <span class="n">save_fit</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>get_go_gene_importances saves the activations, GO importances and further information in a pandas pickle file. We then load these files and visualize the GO importance for a selection of GO terms for all available cell types.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">go_importances</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">cell_types</span> <span class="o">=</span> <span class="p">[]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="nb">sorted</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">])):</span>
    <span class="n">cell_type_mod</span> <span class="o">=</span> <span class="n">cell_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">go_importances</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;./tutorial/calc_go_terms/protein_cell_type_vs_rest_</span><span class="si">{</span><span class="n">cell_type_mod</span><span class="si">}</span><span class="s2">_res_unp_act_mask_False_flip_False_go_terms_evaluated_graph_csv.pkl&quot;</span><span class="p">)[</span><span class="s2">&quot;predictors_label_valid_roc&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;float&quot;</span><span class="p">))</span>
        <span class="n">cell_types</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">cell_type</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">FileNotFoundError</span><span class="p">:</span>
        <span class="k">continue</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">go_terms</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_pickle</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;./tutorial/calc_go_terms/protein_cell_type_vs_rest_</span><span class="si">{</span><span class="n">cell_type_mod</span><span class="si">}</span><span class="s2">_res_unp_act_mask_False_flip_False_go_terms_evaluated_graph_csv.pkl&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="n">go_terms</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">obodag</span><span class="p">[</span><span class="n">go_term</span><span class="p">]</span><span class="o">.</span><span class="n">name</span><span class="si">}</span><span class="s2"> (</span><span class="si">{</span><span class="n">go_term</span><span class="si">}</span><span class="s2">)&quot;</span> <span class="k">if</span> <span class="n">go_term</span> <span class="ow">in</span> <span class="n">obodag</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span> <span class="k">else</span> <span class="n">go_term</span> <span class="k">for</span> <span class="n">go_term</span> <span class="ow">in</span> <span class="n">go_terms</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">go_importances</span><span class="p">)</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">xticklabels</span><span class="o">=</span><span class="n">cell_types</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="n">go_terms</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/7feb4d3078d99e99812fa4653504ef07f1d693acf821fd313736061b8a417d4b.png" src="../_images/7feb4d3078d99e99812fa4653504ef07f1d693acf821fd313736061b8a417d4b.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="s2">&quot;tutorial/calc_go_terms&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We also show how to perform a sensitivity analysis by perturbing features. Here we perturb CD86 (ENSG00000114013) by setting all values to zero. The function get_perturbation_go_gene_importances internally calls get_go_gene_importances, but additionally we need to provide the modality of the features in modality_key and the ENSEMBL-IDs in gene_stable_id_key.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">perturbation_gene_stable_ids</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;ENSG00000114013&quot;</span><span class="p">]</span>

<span class="n">go_importances_perturbation</span> <span class="o">=</span> <span class="n">vae</span><span class="o">.</span><span class="n">get_perturbation_go_gene_importances</span><span class="p">(</span><span class="n">labels_column</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span>
                                               <span class="n">modality_key</span><span class="o">=</span><span class="s2">&quot;modality&quot;</span><span class="p">,</span>
                                               <span class="n">gene_stable_id_key</span><span class="o">=</span><span class="s2">&quot;gene_stable_id&quot;</span><span class="p">,</span>
                                               <span class="n">perturbation_gene_stable_ids</span><span class="o">=</span><span class="n">perturbation_gene_stable_ids</span><span class="p">,</span>
                                               <span class="n">batch_size</span><span class="o">=</span><span class="mi">512</span><span class="p">,</span>
                                               <span class="n">shuffle_set_split</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                               <span class="n">calc_go_terms</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                                               <span class="n">calc_genes</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                                               <span class="n">calc_gene_groups</span><span class="o">=</span><span class="kc">False</span><span class="p">,)</span>
</pre></div>
</div>
</div>
</div>
<p>We can visualize the change of the importance in comparison to the not-perturbed dataset accordingly:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">go_importances_pert</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">cell_type</span> <span class="ow">in</span> <span class="n">cell_types</span><span class="p">:</span>
    <span class="n">cell_type_mod</span> <span class="o">=</span> <span class="n">cell_type</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
    <span class="n">go_importances_pert</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">go_importances_perturbation</span><span class="p">[</span><span class="s2">&quot;ENSG00000114013&quot;</span><span class="p">][</span><span class="s1">&#39;protein&#39;</span><span class="p">][</span><span class="s2">&quot;protein&quot;</span><span class="p">][</span><span class="n">cell_type_mod</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;float&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mi">6</span><span class="p">))</span>

<span class="n">sns</span><span class="o">.</span><span class="n">heatmap</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">go_importances</span><span class="p">)</span> <span class="o">-</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">go_importances_pert</span><span class="p">))</span><span class="o">.</span><span class="n">T</span><span class="p">,</span> <span class="n">xticklabels</span><span class="o">=</span><span class="n">cell_types</span><span class="p">,</span> <span class="n">yticklabels</span><span class="o">=</span><span class="n">go_terms</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/6ca408f168048640ca6be5bb4dd20633e73173c41b34f66604e3f20b839fac68.png" src="../_images/6ca408f168048640ca6be5bb4dd20633e73173c41b34f66604e3f20b839fac68.png" />
</div>
</div>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="mosaic_integration.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Mosaic integration</p>
      </div>
    </a>
    <a class="right-next"
       href="go_specific_covariate_attention.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Interpretability: Inferring GO-specific covariate attention values</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-loading">Data loading</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#data-setup">Data setup</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#model-setup-and-training">Model setup and training</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#networkvi">NetworkVI</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#interpretability">Interpretability</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Lucas Arnoldt, Julius Upmeier zu Belzen, Luis Herrmann, Khue Nguyen, Fabian Theis, Benjamin Wild, Roland Eils
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Lucas Arnoldt, Julius Upmeier zu Belzen, Luis Herrmann, Khue Nguyen, Fabian Theis, Benjamin Wild, Roland Eils.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>